omp/solver/gmres_kernels.cpp

```diff
@@ -53,6 +54,201 @@ namespace omp {
+
+    for (size_type i = 0; i < next_krylov_basis->get_size()[1]; ++i) {
+        if (stop_status[i].has_stopped()) {
+            continue;
+        }
+        for (size_type k = 0; k < iter + 1; ++k) {
+            ValueType hessenberg_iter_entry = zero<ValueType>();
+
+#pragma omp parallel for reduction(add : hessenberg_iter_entry)
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                hessenberg_iter_entry +=
+                    next_krylov_basis->at(j, i) *
+                    krylov_bases->at(j,
+                                     next_krylov_basis->get_size()[1] * k + i);
+            }
+            hessenberg_iter->at(k, i) = hessenberg_iter_entry;
+
+#pragma omp parallel for
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                next_krylov_basis->at(j, i) -=
+                    hessenberg_iter->at(k, i) *
+                    krylov_bases->at(j,
+                                     next_krylov_basis->get_size()[1] * k + i);
+            }
+        }
+        // for i in 1:iter
+        //     hessenberg(iter, i) = next_krylov_basis' * krylov_bases(:, i)
+        //     next_krylov_basis  -= hessenberg(iter, i) * krylov_bases(:, i)
+        // end
+
+        ValueType hessenberg_iter_entry = zero<ValueType>();
+
+#pragma omp parallel for reduction(add : hessenberg_iter_entry)
+        for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+            hessenberg_iter_entry +=
+                next_krylov_basis->at(j, i) * next_krylov_basis->at(j, i);
+        }
+        hessenberg_iter->at(iter + 1, i) = sqrt(hessenberg_iter_entry);
+        // hessenberg(iter, iter + 1) = norm(next_krylov_basis)
+#pragma omp parallel for
+        for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+            next_krylov_basis->at(j, i) /= hessenberg_iter->at(iter + 1, i);
+            krylov_bases->at(j, next_krylov_basis->get_size()[1] * (iter + 1) +
+                                    i) = next_krylov_basis->at(j, i);
+        }
+        // next_krylov_basis /= hessenberg(iter, iter + 1)
+        // krylov_bases(:, iter + 1) = next_krylov_basis
+        // End of arnoldi
+    }
+}
+
+
+template <typename ValueType>
+void calculate_sin_and_cos(matrix::Dense<ValueType> *givens_sin,
+                           matrix::Dense<ValueType> *givens_cos,
+                           matrix::Dense<ValueType> *hessenberg_iter,
+                           size_type iter, const size_type rhs)
+{
+    if (hessenberg_iter->at(iter, rhs) == zero<ValueType>()) {
+        givens_cos->at(iter, rhs) = zero<ValueType>();
+        givens_sin->at(iter, rhs) = one<ValueType>();
+    } else {
+        auto hypotenuse = sqrt(hessenberg_iter->at(iter, rhs) *
+                                   hessenberg_iter->at(iter, rhs) +
+                               hessenberg_iter->at(iter + 1, rhs) *
```

Comment:
Is this alignment correct? It looks a bit weird.

Reply:
The alignment was done by make format. Should I leave it or change it?

Reply:
I just tried to change it manually and run `make format` over it, it changed it again to what is present here.
I am pretty sure it detects that `*` has higher priority than `+`, therefore trying to show the closer binding of the factors with this kind of indentation.

I would recommend just leaving as it is since `clang-format` would change it back (unless you would turn it off for this section, but I don't think it is necessary here).

Reply:
That's fine, if `make format` does it then let's not bother too much.